File: typescript/src/integrations/llamaindex/postprocessor.ts
/**
 * `BaseNodePostprocessor` that compresses retrieved node content.
 *
 * Mirrors Python `CompresrNodePostprocessor`. Query comes from the
 * `QueryBundle` supplied by the query engine; override via `query`.
 */
import type { CompressionClient } from '../../clients/compression.js';
import { getLogger } from '../../logger.js';
import {
  BATCH_LIMIT,
  buildClient,
  DEFAULT_MIN_TOKENS,
  DEFAULT_MODEL,
  DEFAULT_POLICY,
  DEFAULT_RATIO,
  type ErrorPolicy,
  estimateTokens,
} from '../_shared/index.js';

import type { NodeWithScore } from '@llamaindex/core/schema';

type NodeWithScoreType = NodeWithScore;

/**
 * LlamaIndex.TS query input — either the raw query string or an object
 * carrying the query string (matching LlamaIndex.TS' `MessageContent`
 * conventions plus Python-side `QueryBundle.query_str`).
 */
type QueryInput =
  | string
  | { queryStr?: string; query_str?: string }
  | undefined;

export interface CompresrNodePostprocessorOptions {
  apiKey?: string;
  client?: CompressionClient;
  baseUrl?: string;
  compressionModel?: string;
  targetCompressionRatio?: number;
  /**
   * Absolute output token budget per node. When set, overrides
   * `targetCompressionRatio` with `ratio = avg_chunk_tokens / targetToken`.
   * Approximate (chars/4 estimator).
   */
  targetToken?: number;
  minTokens?: number;
  coarse?: boolean;
  onError?: ErrorPolicy;
  /** Override the query supplied by the query engine. */
  query?: string;
}

/**
 * @example
 * ```ts
 * import { CompresrNodePostprocessor } from '@compresr/sdk/integrations/llamaindex';
 * const pp = new CompresrNodePostprocessor({
 *   apiKey: process.env.COMPRESR_API_KEY!,
 *   compressionModel: 'latte_v1',
 * });
 * const queryEngine = index.asQueryEngine({ nodePostprocessors: [pp] });
 * ```
 */
export class CompresrNodePostprocessor {
  private readonly compresr: CompressionClient;
  private readonly compressionModel: string;
  private readonly targetCompressionRatio: number;
  private readonly targetToken: number | undefined;
  private readonly minTokens: number;
  private readonly coarse: boolean | undefined;
  private readonly onError: ErrorPolicy;
  private readonly staticQuery: string | undefined;

  constructor(options: CompresrNodePostprocessorOptions) {
    const {
      apiKey,
      client,
      baseUrl,
      compressionModel = DEFAULT_MODEL,
      targetCompressionRatio = DEFAULT_RATIO,
      targetToken,
      minTokens = DEFAULT_MIN_TOKENS,
      coarse,
      onError = DEFAULT_POLICY,
      query,
    } = options;

    this.compresr =
      client ?? buildClient({ apiKey, baseUrl, caller: 'CompresrNodePostprocessor' });
    this.compressionModel = compressionModel;
    this.targetCompressionRatio = targetCompressionRatio;
    this.targetToken = targetToken;
    this.minTokens = minTokens;
    this.coarse = coarse;
    this.onError = onError;
    this.staticQuery = query;
  }

  private effectiveRatio(chunk: readonly string[]): number {
    if (!this.targetToken || this.targetToken <= 0) return this.targetCompressionRatio;
    const avg = Math.max(
      1,
      chunk.reduce((acc, c) => acc + estimateTokens(c), 0) / Math.max(1, chunk.length)
    );
    return Math.max(avg / this.targetToken, 1.0);
  }

  private resolveQuery(queryBundle?: QueryInput | string): string | undefined {
    if (this.compressionModel !== 'latte_v1') return undefined;
    if (typeof this.staticQuery === 'string' && this.staticQuery.trim()) {
      return this.staticQuery;
    }
    if (typeof queryBundle === 'string' && queryBundle.trim()) {
      return queryBundle;
    }
    if (queryBundle && typeof queryBundle === 'object') {
      const q = (queryBundle).queryStr ??
        (queryBundle as { query_str?: string }).query_str;
      if (typeof q === 'string' && q.trim()) return q;
    }
    return undefined;
  }

  private partition(
    nodes: readonly NodeWithScoreType[]
  ): { indices: number[]; contents: string[] } {
    const indices: number[] = [];
    const contents: string[] = [];
    nodes.forEach((n, i) => {
      const text = getNodeText(n);
      if (typeof text === 'string' && estimateTokens(text) >= this.minTokens) {
        indices.push(i);
        contents.push(text);
      }
    });
    return { indices, contents };
  }

  /** LlamaIndex.TS `BaseNodePostprocessor.postprocessNodes` signature. */
  async postprocessNodes(
    nodes: readonly NodeWithScoreType[],
    queryBundle?: QueryInput | string
  ): Promise<NodeWithScoreType[]> {
    if (nodes.length === 0) return [...nodes];

    const { indices, contents } = this.partition(nodes);
    if (contents.length === 0) return [...nodes];

    const query = this.resolveQuery(queryBundle);
    if (this.compressionModel === 'latte_v1' && !query) {
      getLogger().warn(
        'CompresrNodePostprocessor: latte_v1 needs a query; none found. Passthrough.'
      );
      return [...nodes];
    }

    const out = nodes.map(cloneNodeWithScore);

    for (let start = 0; start < contents.length; start += BATCH_LIMIT) {
      const chunk = contents.slice(start, start + BATCH_LIMIT);
      let results: { compressed_context: string }[];
      try {
        const resp = await this.compresr.compressBatch({
          contexts: chunk,
          ...(query !== undefined ? { queries: query } : {}),
          compressionModelName: this.compressionModel,
          targetCompressionRatio: this.effectiveRatio(chunk),
          ...(this.coarse !== undefined ? { coarse: this.coarse } : {}),
        });
        results = (resp.data?.results ?? []);
      } catch (exc) {
        if (this.onError === 'raise') throw exc;
        getLogger().warn(
          `postprocessor batch failed (${stringifyError(exc)}); passthrough ${chunk.length} nodes.`
        );
        continue;
      }

      results.slice(0, chunk.length).forEach((item, offset) => {
        const globalIdx = indices[start + offset];
        const newText = item?.compressed_context;
        if (typeof newText === 'string' && newText && globalIdx !== undefined) {
          const node = out[globalIdx];
          if (node !== undefined) {
            setNodeText(node, newText);
          }
        }
      });
    }

    return out;
  }
}


function getNodeText(nws: NodeWithScoreType): string | undefined {
  const node = (nws as { node?: unknown }).node;
  if (!node || typeof node !== 'object') return undefined;
  const n = node as { getContent?: () => string; text?: string };
  if (typeof n.getContent === 'function') {
    try {
      const t = n.getContent();
      if (typeof t === 'string') return t;
    } catch {
      /* fall through */
    }
  }
  if (typeof n.text === 'string') return n.text;
  return undefined;
}

function cloneNodeWithScore(nws: NodeWithScoreType): NodeWithScoreType {
  return { ...(nws as object) } as NodeWithScoreType;
}

function setNodeText(nws: NodeWithScoreType, newText: string): void {
  const node = (nws as { node?: unknown }).node;
  if (!node || typeof node !== 'object') return;
  const n = node as { setContent?: (t: string) => void; text?: string };
  if (typeof n.setContent === 'function') {
    try {
      n.setContent(newText);
      return;
    } catch {
      /* fall through */
    }
  }
  try {
    n.text = newText;
    return;
  } catch {
    /* fall through */
  }
  const md = (node as { metadata?: Record<string, unknown> }).metadata;
  if (md && typeof md === 'object') {
    md['compresr_compressed'] = newText;
  }
  getLogger().warn(
    'node has no writable text field; compressed text stored in metadata.compresr_compressed but the synthesizer will see the original text.'
  );
}

function stringifyError(exc: unknown): string {
  return exc instanceof Error ? exc.message : String(exc);
}

File: python/compresr/integrations/llamaindex/postprocessor.py
"""``BaseNodePostprocessor`` that compresses retrieved node content.

The query is supplied by the query engine via ``QueryBundle.query_str``;
``query=...`` overrides it (rarely needed).

Example::

    from llama_index.core import VectorStoreIndex
    from compresr.integrations.llamaindex import CompresrNodePostprocessor

    pp = CompresrNodePostprocessor(api_key=os.environ["COMPRESR_API_KEY"])
    query_engine = index.as_query_engine(node_postprocessors=[pp])
"""

from __future__ import annotations

import logging
from typing import Any, List, Optional

from .._shared import (
    BATCH_LIMIT,
    DEFAULT_MIN_TOKENS,
    DEFAULT_MODEL,
    DEFAULT_POLICY,
    DEFAULT_RATIO,
    ErrorPolicy,
    build_client,
    estimate_tokens,
)

logger = logging.getLogger(__name__)

try:
    from llama_index.core.bridge.pydantic import Field  # type: ignore[import-not-found]
    from llama_index.core.postprocessor.types import (  # type: ignore[import-not-found]
        BaseNodePostprocessor,
    )
    from llama_index.core.schema import NodeWithScore, QueryBundle  # type: ignore[import-not-found]
except ImportError as exc:  # pragma: no cover
    raise ImportError(
        "CompresrNodePostprocessor requires llama-index-core. "
        "Install with: pip install compresr[llamaindex]"
    ) from exc


class CompresrNodePostprocessor(BaseNodePostprocessor):
    """Compress retrieved nodes with query-aware compression.

    Eligible nodes (content >= ``min_tokens``) are compressed in a single
    batch call per ``BATCH_LIMIT`` slice. Compressed text replaces
    ``node.text`` on a *copy* — the original index is untouched.
    """

    api_key: Optional[str] = Field(default=None)
    base_url: Optional[str] = Field(default=None)
    compression_model: str = Field(default=DEFAULT_MODEL)
    target_compression_ratio: float = Field(default=DEFAULT_RATIO)
    target_token: Optional[int] = Field(
        default=None,
        description=(
            "Absolute output token budget per node. When set, overrides "
            "target_compression_ratio with `ratio = max(estimate_tokens(ctx) / "
            "target_token, 1.0)`. Approximate (chars/4 estimator)."
        ),
    )
    min_tokens: int = Field(default=DEFAULT_MIN_TOKENS)
    coarse: Optional[bool] = Field(default=None)
    on_error: ErrorPolicy = Field(default=DEFAULT_POLICY)
    query: Optional[str] = Field(
        default=None,
        description="Override the query passed by the query engine.",
    )
    client: Any = Field(default=None, exclude=True)

    model_config = {"arbitrary_types_allowed": True}

    @classmethod
    def class_name(cls) -> str:
        return "CompresrNodePostprocessor"

    def _get_client(self) -> Any:
        if self.client is not None:
            return self.client
        self.client = build_client(
            api_key=self.api_key,
            base_url=self.base_url,
            caller="CompresrNodePostprocessor",
        )
        return self.client

    def _resolve_query(self, query_bundle: Optional["QueryBundle"]) -> Optional[str]:
        if isinstance(self.query, str) and self.query.strip():
            return self.query
        if query_bundle is not None:
            q = getattr(query_bundle, "query_str", None)
            if isinstance(q, str) and q.strip():
                return q
        return None

    def _partition(self, nodes: List["NodeWithScore"]) -> tuple[list[int], list[str]]:
        indices: list[int] = []
        contents: list[str] = []
        for i, n in enumerate(nodes):
            text = n.node.get_content()
            if isinstance(text, str) and estimate_tokens(text) >= self.min_tokens:
                indices.append(i)
                contents.append(text)
        return indices, contents

    def _apply_results(
        self,
        nodes: List["NodeWithScore"],
        indices: list[int],
        results: list,
        slice_start: int,
        chunk_len: int,
    ) -> None:
        for offset, item in enumerate(results[:chunk_len]):
            global_idx = indices[slice_start + offset]
            new = getattr(item, "compressed_context", None)
            if isinstance(new, str) and new:
                _set_node_text(nodes[global_idx], new)

    def _effective_ratio(self, chunk: List[str]) -> float:
        """Per-chunk compression ratio.

        Backend accepts only ``target_compression_ratio`` on the wire. When
        ``target_token`` is set, translate it client-side using the average
        per-context token estimate for the chunk."""
        if not self.target_token or self.target_token <= 0:
            return self.target_compression_ratio
        avg = max(1.0, sum(estimate_tokens(c) for c in chunk) / max(1, len(chunk)))
        return max(avg / self.target_token, 1.0)

    def _postprocess_nodes(
        self,
        nodes: List["NodeWithScore"],
        query_bundle: Optional["QueryBundle"] = None,
    ) -> List["NodeWithScore"]:
        if not nodes:
            return nodes

        indices, contents = self._partition(nodes)
        if not contents:
            return nodes

        query = self._resolve_query(query_bundle)
        if not query:
            logger.warning(
                "CompresrNodePostprocessor: no query resolved; passthrough. "
                "Provide via QueryBundle.query_str or `query=` on the postprocessor."
            )
            return nodes

        client = self._get_client()
        out = [_clone_node_with_score(n) for n in nodes]

        for start in range(0, len(contents), BATCH_LIMIT):
            chunk = contents[start : start + BATCH_LIMIT]
            ratio = self._effective_ratio(chunk)
            try:
                resp = client.compress_batch(
                    contexts=chunk,
                    queries=query,
                    compression_model_name=self.compression_model,
                    target_compression_ratio=ratio,
                    coarse=self.coarse,
                )
            except Exception as exc:  # noqa: BLE001
                if self.on_error == "raise":
                    raise
                logger.warning(
                    "compresr: postprocessor batch failed (%s); passthrough %d nodes.",
                    exc,
                    len(chunk),
                )
                continue

            results = (resp.data.results if resp.data else None) or []
            self._apply_results(out, indices, results, start, len(chunk))

        return out


def _clone_node_with_score(nws: "NodeWithScore") -> "NodeWithScore":
    if hasattr(nws.node, "model_copy"):
        new_node = nws.node.model_copy()
    elif hasattr(nws.node, "copy"):
        new_node = nws.node.copy()
    else:
        logger.warning(
            "CompresrNodePostprocessor: node type %s lacks model_copy/copy; "
            "downstream mutations will affect the original.",
            type(nws.node).__name__,
        )
        new_node = nws.node
    return NodeWithScore(node=new_node, score=nws.score)


def _set_node_text(nws: "NodeWithScore", new_text: str) -> None:
    node = nws.node
    if hasattr(node, "set_content"):
        try:
            node.set_content(new_text)
            return
        except Exception:
            pass
    if hasattr(node, "text"):
        try:
            node.text = new_text
            return
        except Exception:
            pass
    md = getattr(node, "metadata", None)
    if isinstance(md, dict):
        md["compresr_compressed"] = new_text
    logger.warning(
        "CompresrNodePostprocessor: node type %s has no writable text field; "
        "compressed text stored in metadata[compresr_compressed] but the "
        "synthesizer will see the original text via get_content().",
        type(node).__name__,
    )
